### ISSUE TITLE
[Proposal] Non-static SerializationManager changes

### ISSUE THREAD

**Issue by ReubenBond** [dotnet/orleans#2547]
In order to complete the work required for non-static grain clients (#467), it would be very helpful to make `SerializationManager` also non-static so that we can deserialize grain references correctly.

`GrainReferences` need to have access to services provided by Orleans in order to function (for example, they need to make remote calls). Therefore, I think it's a good idea to change the way that serialization and deserialization work. I'm proposing these changes and want your feedback:

* Serialization methods accept an `ISerializationContext`/`IDeserializationContext`/`ICopyContext` in their signature. The methods will all look something like this:
```C#
public interface IExternalSerializer
{
    void Initialize(Logger logger);
    bool IsSupportedType(Type itemType);
    object DeepCopy(object source, ICopyContext context);
    void Serialize(object item, ISerializationContext context, Type expectedType);
    object Deserialize(Type expectedType, IDeserializationContext context);
}
```

`SerializationContext` & `DerserializationContext`:
* Add a `Stream` property which includes the `BinaryTokenStreamWriter`/`BinaryTokenStreamReader`.
* Add a `ServiceProvider` property of type `IServiceProvider`.
* Add a `SerializationManager` property for convenient access to `SerializationManager` (could also be accessed via DI).
* Are pooled rather than being thread-static.

The implications of this are:
* `SerializationManager` can be made non-static
* Serializers can access services provided by Orleans (or the user, since it's the same container)
* Serializers will need to be adjusted to satisfy the new signature (i.e, this is a **breaking change**), but I will create Roslyn Analyzers to ensure that breakages aren't hidden or inexplicable: ![Roslyn Analyzer Prototype](http://i.imgur.com/koq7g2t.gif)

Alternatives:
* Keep method signatures the same and pass those context objects inside of `BinaryTokenStreamReader` and `BinaryTokenStreamWriter`. I think this is messy and myopic.
* Add an `IOnDeserialized` interface which accepts an `IServiceProvider`. I played with this. It will help for non-static grain clients, but not with making `SerializationManager` non-static.
* Others?

Thoughts?

**Comment by jdom** [comment]
I agree with adding a context, and avoid thread-statics, static singletons, and I would even avoid the IServiceProvider for getting the SerializationManager (while doing serialization, the serialization manager is a very important piece of the context and shouldn't be hidden by the service provider, which should be a last resort for edge cases).

Since we are doing breaking changes, how about becoming a little bit more functional, and instead of having the user define 3 different static methods that must be attributed correctly, instead require a single method that returns the `SerializerMethods` struct that contains all 3 methods. It's a little less fragile, as it is not possible to define the serializer but not the deserializer. In theory Copier can be defined independently though, but it's probably ok to use SerializerMethods regardless (and set null or an default copier). This will make the breaking change a little less subtle too.

**Comment by ReubenBond** [comment]
@jdom thanks for the input!

Would you have something like this?
```C#
public static void RegisterSerializers(ISerializerCollection serializers)
{
  serializers.Register(
    typeof(MyType),
    (obj, ctx, expected) => Serialize(obj, ctx, expected), ...etc...)
}
```

In that case, we could also move those things to the (non-existant) builder with a helper method and resolve them via the service provider.

**Comment by ReubenBond** [comment]
The implementation of this has been merged #2558.
